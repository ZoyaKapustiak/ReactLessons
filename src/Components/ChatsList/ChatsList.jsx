import React from "react";
import { useState } from "react";
import { Link} from 'react-router-dom';
import styles from './ChatsList.module.css'
import {TextField, List, ListItem, Button, Box} from '@mui/material';
import { useDispatch } from 'react-redux'
import { addChat } from '../../store/messages/actions'
import { push, set, remove } from 'firebase/database'
import { messagesRef, getChatById, getMessageListById } from "../../services/firebase";

export function ChatsList({messageDB, chats}) {
  const [chat, setChat] = useState('')
  const dispatch = useDispatch()

  const handleSubmit = (e) => {
    e.preventDefault() 
    dispatch(addChat(chat))
    
    set(messagesRef, {
      ...messageDB,
      [chat]: {
        name: chat
      }
    })

    push(getMessageListById(chat), {
      text: 'Chat has been created',
      author: 'Admin'
    });

    setChat('')
  }

  const handleDeleteChat = (chatId) => {
    remove(getChatById(chatId));
  };

  return (
  <>
    <h1> ChatList</h1> 
    <List className={styles.listBlock}>
      {chats.map((value, ind) => (
      <ListItem
        key={ind}
        disableGutters
        >
        <Link 
          key={ind}
          to={`/chats/${value.name}`}>
          {value.name}
        </Link>
        <Button 
        onClick={() => dispatch(handleDeleteChat(value.name))}
        >X</Button>
      </ListItem>
      ))}
    </List>
    <Box
      onSubmit={handleSubmit}
      component="form"
      sx={{
        '& .MuiTextField-root': { m: 1, width: '25ch' },
      }}
      noValidate
      autoComplete="off">
        <TextField
          id="outlined-multiline-flexible"
          label="Multiline"
          multiline
          maxRows={4}
          value={chat}
          onChange={(e) => setChat(e.target.value)}
          />
        <br/>
        <Button type='submit' variant="outlined" >Create Chat</Button>
    </Box>  
  </>
)
}