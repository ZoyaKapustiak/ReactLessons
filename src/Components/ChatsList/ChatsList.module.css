.listBlock {
  max-width: 360px;
  width: 100%;
}