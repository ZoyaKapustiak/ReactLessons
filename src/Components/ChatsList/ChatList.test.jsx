import React from "react";
import { render, screen } from "@testing-library/react";
import '@testing-library/jest-dom'

import { ChatsList } from "./ChatsList";

import * as reduxHooks from 'react-redux'
import * as actions from ''

jest.mock('react-redux')
const mockUseDispatch = jest.spyOn(reduxHooks, 'useDispatch')
describe('testing ChatList', () => {
  it('should create ChatList', () => {
    mockUseDispatch.mockReturnValue(jest.fn)
    const component = render(<ChatsList messageDB={{}} chats={[]} />)
    expect(component).toMatchSnapshot()
  })

  it('should dispatch action', () => {
    const dispatch = jest.fn;
    mockUseDispatch.mockReturnValue(dispatch)
    const messagesRef = {}
    render(
      <ChatsList messageDB={{}} chats={[]} />
      )
    const buttons = screen.getByRole('button')
    
  })

})