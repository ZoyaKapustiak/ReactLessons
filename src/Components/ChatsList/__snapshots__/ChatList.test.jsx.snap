// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ChatList should create ChatList 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
         ChatList
      </h1>
      <ul
        class="MuiList-root MuiList-padding listBlock css-h4y409-MuiList-root"
      />
      <form
        autocomplete="off"
        class="MuiBox-root css-13r4j3i"
        novalidate=""
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="outlined-multiline-flexible"
            id="outlined-multiline-flexible-label"
          >
            Multiline
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <textarea
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              id="outlined-multiline-flexible"
              style="height: 0px; overflow: hidden;"
            />
            <textarea
              aria-hidden="true"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              readonly=""
              style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
              tabindex="-1"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Multiline
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <br />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Create Chat
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <h1>
       ChatList
    </h1>
    <ul
      class="MuiList-root MuiList-padding listBlock css-h4y409-MuiList-root"
    />
    <form
      autocomplete="off"
      class="MuiBox-root css-13r4j3i"
      novalidate=""
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="outlined-multiline-flexible"
          id="outlined-multiline-flexible-label"
        >
          Multiline
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            id="outlined-multiline-flexible"
            style="height: 0px; overflow: hidden;"
          />
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Multiline
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <br />
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Create Chat
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
